### PR TITLE
Add nested context descriptions to lock documents

### PR DESCRIPTION
**Pull request by apeteri**
This is required to track re-entrant locks that go more than two levels deep.

### CODE REVIEW

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/OperationLock.java [hunk 1]
```diff
@@ -33,8 +34,8 @@ public class OperationLock extends AbstractOperationLock {
 	 * @param id the lock identifier
 	 * @param target the lock target (may not be {@code null})
 	 */
-	public OperationLock(final int id, final Lockable target) {
-		super(id, target);
+	public OperationLock(final int id, final Date creationDate, final Lockable target) {
```
Comment (nagyo) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/OperationLock.java:37]: Wouldn't it be better to use a LocalDate type here?

**Review by nagyo — APPROVED** [review]
RSLGTM! with a note

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/OperationLock.java (hunk 1 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/OperationLock.java:37]: Date includes the time aspect as well, not just a date, so LocalDateTime would be better, but that requires a larger refactor. Keep it as is for now.

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java [hunk 2]
```diff
@@ -140,26 +139,47 @@ public Builder branchPath(final String branchPath) {
...
 	private final String repositoryId;
 	private final String branchPath;
-	
-	private DatastoreLockIndexEntry(final String id, final String userId, final String description, final String parentDescription, final String repositoryId, final String branchPath) {
+	private final Long timestamp;
+	private final List<String> contexts;
+
+	// Only left in for backwards compatibility reasons
+	@Deprecated
+	private final String description = null;
```
Comment (cmark) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java:177]: We currently don't throw away stuck lock entries during restart as we don't know which machine initiated it (yet), therefore there might be a case when there is a lock entry that uses the old format and the system would try to deserialize it but there is not setter in the JSON builder. I'll fix this soon.

**Review by cmark — CHANGES_REQUESTED** [review]

**Review by cmark — APPROVED** [review]

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java (hunk 2 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java:177]: Fixed by [13be6a6](https://github.com/b2ihealthcare/snow-owl/pull/1353/commits/13be6a655c924a0bb41a69a6912a2c651df1cc7d)

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java [hunk 3]
```diff
@@ -143,6 +144,20 @@ public Builder timestamp(final Long timestamp) {
...
+		Builder description(final String description) {
+			// lock documents with description value should propagate their values to contexts
+			addContext(description);
+			return this;
+		}
+		
+		@JsonSetter
+		Builder parentDescription(final String parentDescription) {
+			// lock documents with a parentDescription value should propagate their values to contexts
+			addContext(parentDescription);
```
Comment (apeteri) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/locks/DatastoreLockIndexEntry.java:158]: `parentDescription` goes first in the list, so a `contexts.add(0, parentDescription)` equivalent could be used here. Although whether this is required or not will also depend on the order of the two setter calls.

**Pull request merged**